Balgillo/sev none

balgillo opened the pull request:
Preparing sarif-tools v3.

This started out as a fix for #47 but it's led to a refactoring of the code for grouping issues for:

- Consistent issue sorting and grouping
- Consistent handling of `none` severity
- Consistent treatment of issues with the same issue code but distinct descriptions

debonte commented on sarif/issues_report.py:
```diff
@@ -0,0 +1,146 @@
+            self._sort_record_lists()
+        return self._sev_to_sorted_keys.get(severity, {})
+
+    def get_issue_type_histogram(self, severity: str) -> Dict[str, List[dict]]:
```
The name `get_issue_type_histogram` seems out of place compared with the others, especially the similar `issue_type_count_for_severity`. Maybe it should be renamed to `get_issue_type_counts_grouped_by_type`? I realize that's a bit long, but it shares structure/terminology with both `get_issues_grouped_by_type` and `issue_type_count_for_severity`.

balgillo replied:
I think histogram is a more expressive way to express it, otherwise it's confusing whether we're counting the issues or the types.  Renamed now to `get_issue_type_histogram_for_severity` for improved consistency with other methods.

debonte reviewed (commented):
Obviously I'm unfamiliar with the code base, haven't used the tool, etc., so take my feedback with a grain of salt. :)

balgillo commented on sarif/sarif_file_utils.py:
```diff
@@ -5,8 +5,40 @@
+            description = description[: description.index("\n")]
+        description = description.strip()
+    if description:
+        if len(description) > 120:
```
Thanks for flagging!  I've made it more systematic about keeping to a length budget of 120.  This also led me to a bug which was losing the issue code when finding the common description prefix - now fixed :-)

balgillo commented on sarif/issues_report.py:
```diff
@@ -0,0 +1,146 @@
+
+    def add_record(self, record: dict):
+        """Append record to list for severity - no sorting."""
+        self._sev_to_records.setdefault(record["Severity"], []).append(record)
```
It's not necessary if the input files have stuck to the severities allowed by the SARIF format, but if not this will prevent a crash

balgillo commented on sarif/issues_report.py:
```diff
@@ -0,0 +1,146 @@
+            for record in issues:
+                code = record["Code"]
+                key = combine_code_and_description(record)
+                key_and_count = code_to_key_and_count.get(code, None)
```
Thanks, removed the explicit default value

balgillo commented on sarif/issues_report.py:
```diff
@@ -0,0 +1,146 @@
+        included at the end if there are any records with severity "none".
+        """
+        return (
+            SARIF_SEVERITIES_WITH_NONE
```
Historically, the tools have always included error, warning and none regardless of counts.  There's an aspect of user education here as it immediately shows what the severity levels are, and also the satisfaction and clarity of seeing "errors: 0" rather than the section just disappearing.

But now we're adding "none" severity, we've deviated from the fixed list a bit, because it's a weird severity that most tools won't produce, so we don't want to include it by default.  I think the arguments for always including the other three still stand though.

I'll add some comments to explain

balgillo commented on sarif/issues_report.py:
```diff
@@ -0,0 +1,146 @@
+                records.sort(key=record_sort_key)
+        self._records_have_been_sorted = True
+
+    def issue_count_for_severity(self, severity: str) -> int:
```
Thanks for pointing out the inconsistent naming - it's great to have a fresh pair of eyes on this!  On balance I like the `_for_severity` suffix for clarity.  So adding `_for_severity` and `get_` consistently:
`issue_count_for_severity` -> `get_issue_count_for_severity`
`issue_type_count_for_severity` -> `get_issue_type_count_for_severity`
`get_severities`
`get_issues_grouped_by_type` -> `get_issues_grouped_by_type_for_severity`
`get_issue_type_histogram` -> `get_issue_type_histogram_for_severity`
`get_issues` -> `get_issues_for_severity`

balgillo commented on CHANGELOG.md:
```diff
@@ -5,11 +5,26 @@ All notable changes to this project will be documented in this file.
+
+- Changed Python API to use new IssueReport type for issue grouping and sorting:
+  - `SarifFileSet` now has a `get_report() method
+  - `s.get_result_count_by_severity()` replaced by `s.get_report().get_issue_type_histogram(severity)`
+  - `s.get_result_count_by_severity()` replaced by `s.get_report().issue_count_for_severity(severity)`
+  - `s.get_records_grouped_by_severity()` replaced by `s.get_report().get_issues(severity)`
```
Thanks for pointing out the inconsistent naming.  I've improved it - see other comment reply

balgillo commented on sarif/operations/summary_op.py:
```diff
@@ -59,12 +58,12 @@ def _generate_summary(input_files: SarifFileSet) -> List[str]:
+    report = input_files.get_report()
+    for severity in report.get_severities():
+        result_count = report.issue_count_for_severity(severity)
+        issue_type_histogram = report.get_issue_type_histogram(severity)
```
`get_issue_type_histogram` has issue counts as the values, `get_issues_grouped_by_type` has the list of issues.  So `get_issue_type_histogram` is a convenience method to avoid the need to measure the list lengths, and it's all we need for this summary which just gives issue counts.

balgillo commented:
> Obviously I'm unfamiliar with the code base, haven't used the tool, etc., so take my feedback with a grain of salt. :)

@debonte Huge thanks for reviewing - we now have more consistent naming and I've fixed a bug that I introduced with these changes.  Let me know whether you have any further comments or approve for merge.
